itensor/itdata/qdense.cc

```diff
@@ -444,7 +483,78 @@ doTask(PlusEQ const& P,
     {
     if(B.store.size() == 0) return;
 
-    if(isReal(A) && isCplx(B))
+    //
+    // If B has blocks that A doesn't have,
+    // then we need to widen the storage of A
+    //
+
+    auto r = order(P.is1());
+
+    if(r == 0)
+        {
+        if(isReal(A) && isCplx(B))
+            {
+            auto *nA = m.makeNewData<QDenseCplx>(A.offsets,A.begin(),A.end());
+            add(P,*nA,B);
+            }
+        else
+            {
+            auto *mA = m.modifyData(A);
+            add(P,*mA,B);
+            }
+        return;
+        }
+
+    auto Ablock = Block(r,0),
+         Bblock = Block(r,0);
+
+    // Store the blocks of the output
+    // TODO: optimize this to merge and remove repeats
+    //       at the same time, assuming A.offsets and 
+    //       B.offsets are already sorted
+    auto Cblocks = Blocks();
+
+    for(auto& aio : A.offsets)
+        {
+        Ablock = aio.block;
```

Comment:
Assignments like this turned out to be surprisingly expensive when performed many times, this is a reminder to just use `aio.block` directly instead of assigning to `Ablock`.

Reply:
These are great finds. What are the main methods you’re using to profile? (Just so I can know what’s working the best.)

Reply:
We are using gperftools, Nils has a good set of instructions on how to use it from the C++ bootcamp:

https://github.com/Wentzell/Cpp_and_Tools_Bootcamp/blob/master/Tools/Profiling/rng_example/commands

It really helped us identify the bottlenecks. One small thing is that it is important to run with a single MKL thread (`export OMP_NUM_THREADS = 1`), otherwise you get a lot of extraneous information about the threads in the profile data.

Reply:
Interesting. Glad I asked because I had tried gperftools a while back and was not that impressed, and one reason is that I kepy seeing all of these things about threading in the data. So I needed to know that trick at the time!

Reply:
I'm surprised by the way about the `Ablock = aio.block` line above being a problem. Aren't both of those variables just integers? I would think that (a) that would be very cheap to just copy an integer and (b) the compiler could likely optimize that assignment away. 

But if changing that line does give a measureable speedup then you can't argue with the results! Optimizations can be non-intuitive.

Reply:
In the current design, `aio.block` is an `IntArray` of the cartesian coordinates of the block. `Ablock` is also an `IntArray` (in the code, I made `Block` a typedef for `IntArray`). So the cost was assigning `IntArray`s to each other.

Reply:
Great - makes more sense about the cost of that!

Reply:
Still surprising how much of a cost they were, since they are very small. We definitely wouldn't have caught it without the profiler.